src/ck3/data/prov_terrain.rs

```diff
@@ -0,0 +1,189 @@
+use std::path::PathBuf;
+
+use fnv::FnvHashMap;
+
+use crate::block::Block;
+use crate::everything::Everything;
+use crate::fileset::{FileEntry, FileHandler};
+use crate::helpers::dup_error;
+use crate::item::Item;
+use crate::pdxfile::PdxFile;
+use crate::report::{warn, ErrorKey};
+use crate::token::{Loc, Token};
+use crate::validator::Validator;
+use crate::Severity;
+
+use super::provinces::ProvId;
+
+const DEFAULT_TERRAINS: &[&str] = &["default_land", "default_sea", "default_coastal_sea"];
+
+#[derive(Clone, Debug, Default)]
+pub struct ProvinceTerrains {
+    provinces: FnvHashMap<ProvId, ProvinceTerrain>,
+    file_loc: Option<Loc>,
+    defaults: Box<[Option<Token>; DEFAULT_TERRAINS.len()]>,
+}
+
+impl ProvinceTerrains {
+    fn load_item(&mut self, id: ProvId, key: Token, value: Token) {
+        if let Some(province) = self.provinces.get_mut(&id) {
+            if province.key.loc.kind >= key.loc.kind {
+                dup_error(&key, &province.key, "province");
+            }
+            province.terrain = value;
+        } else {
+            self.provinces.insert(id, ProvinceTerrain::new(key, value));
+        }
+    }
+
+    pub fn validate(&self, data: &Everything) {
+        for (provid, item) in &self.provinces {
+            item.validate(*provid, data);
+        }
+
+        for (name, terrains) in DEFAULT_TERRAINS.iter().zip(&*self.defaults) {
+            if let Some(terrain) = terrains {
+                data.verify_exists(Item::Terrain, terrain);
+            } else {
+                let msg = format!("missing default terrain: {}", name);
+                warn(ErrorKey::Validation)
+                    .msg(msg)
+                    // SAFETY: `file_loc` has been set in `handle_file`.
```

Comment:
Turns out this panics if there were no province terrain files at all.

It runs into a general problem with tiger, which is that there's not always a good Loc to use for a missing item.

Maybe a synthetic FileEntry for the whole directory would be a good default. FileEntry can be used as an errorloc and can contain arbitrary paths, unlike Loc which has a PathTableIndex.

Reply:
Yes that's the problem. Of course there is the FileEntry from `handle_file` but that faces the same problem right now. I wonder if it is possible to have a Static Loc that can be used for all missing items.